source/shaders/ibl_filtering.frag

```diff
@@ -149,207 +152,251 @@ float D_Charlie(float sheenRoughness, float NdotH)
+    return clamp(1.0 / (4.0 * (NdotL + NdotV - NdotL * NdotV)), 0.0, 1.0);
+}
+
+// Mipmap Filtered Samples (GPU Gems 3, 20.4)
+// https://developer.nvidia.com/gpugems/gpugems3/part-iii-rendering/chapter-20-gpu-based-importance-sampling
+float computeLod(float pdf)
+{
+    // IBL Baker (Matt Davidson)
+    // https://github.com/derkreature/IBLBaker/blob/65d244546d2e79dd8df18a28efdabcf1f2eb7717/data/shadersD3D11/IblImportanceSamplingDiffuse.fx#L215
+    float solidAngleTexel = 4.0 * MATH_PI / (6.0 * float(u_width) * float(u_sampleCount));
+    float solidAngleSample = 1.0 / (float(u_sampleCount) * pdf);
+    float lod = 0.5 * log2(solidAngleSample / solidAngleTexel);
+
+    lod += u_lodBias;
+
+    return lod;
 }
 
 vec3 filterColor(vec3 N)
 {
-	//return  textureLod(uCubeMap, N, 3.0).rgb;
-	vec4 color = vec4(0.f);
-	float solidAngleTexel = 4.0 * MATH_PI / (6.0 * float(u_width) * float(u_width));
-
-	for(int i = 0; i < u_sampleCount; ++i)
-	{
-		vec3 H = getSampleVector(i, N, u_roughness);
-
-		// Note: reflect takes incident vector.
-		// Note: N = V
-		vec3 V = N;
-
-		vec3 L = normalize(reflect(-V, H));
-
-		float NdotL = dot(N, L);
-
-		if (NdotL > 0.0)
-		{
-			float lod = 0.0;
-
-			if (u_roughness > 0.0 ||u_distribution == cLambertian)
-			{
-				// Mipmap Filtered Samples
-				// see https://github.com/derkreature/IBLBaker
-				// see https://developer.nvidia.com/gpugems/GPUGems3/gpugems3_ch20.html
-				float pdf = PDF(V, H, N, L, u_roughness );
-
-				float solidAngleSample = 1.0 / (float(u_sampleCount) * pdf);
-
-				lod = 0.5 * log2(solidAngleSample / solidAngleTexel);
-				lod += u_lodBias;
-			}
-
-			if(u_distribution == cLambertian)
-			{
-				color += vec4(textureLod(uCubeMap, H, lod).rgb, 1.0);
-			}
-			else
-			{
-				color += vec4(textureLod(uCubeMap, L, lod).rgb * NdotL, NdotL);
-			}
-		}
-	}
-
-	if(color.w == 0.f)
-	{
-		return color.rgb;
-	}
-
-	return color.rgb / color.w;
+    //return  textureLod(uCubeMap, N, 3.0).rgb;
+    vec4 color = vec4(0.f);
+
+    for(int i = 0; i < u_sampleCount; ++i)
+    {
+        vec4 importanceSample = getImportanceSample(i, N, u_roughness);
+
+        vec3 H = vec3(importanceSample.xyz);
+        float pdf = importanceSample.w;
+
+        // mipmap filtered samples (GPU Gems 3, 20.4)
+        float lod = computeLod(pdf);
+
+        if(u_distribution == cLambertian)
+        {
+            float NdotH = clamp(dot(N, H), 0.0, 1.0);
+
+            // sample lambertian at a lower resolution to avoid fireflies
+            vec3 lambertian = textureLod(uCubeMap, H, lod).rgb;
+
+            //// the below operations cancel each other out
+            // lambertian *= NdotH; // lamberts law
+            // lambertian /= pdf; // invert bias from importance sampling
+            // lambertian /= MATH_PI; // convert irradiance to radiance https://seblagarde.wordpress.com/2012/01/08/pi-or-not-to-pi-in-game-lighting-equation/
+
+            color += vec4(lambertian, 1.0);
+            continue;
```

Comment:
Is this a special case for lambertian distribution? - It should be evident that we are filtering either lambertian or ggx